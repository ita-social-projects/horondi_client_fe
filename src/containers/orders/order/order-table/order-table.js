import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Table, TableBody, TableCell, TableHead, TableRow } from '@material-ui/core';
import { useStyles } from './order-table.styles';

import CartItem from '../../cart/cart-item';
import ConfirmDialog from '../../../../components/confirm-dialog';
import PageTitle from '../../../../components/page-title/page-title';

const OrderTable = ({ items, user, cartOperations, promoCode, certificateData }) => {
  const { t, i18n } = useTranslation();
  const language = i18n.language === 'ua' ? 0 : 1;
  const styles = useStyles();
  const [removeOneModalVisibility, setRemoveOneModalVisibility] = useState(false);
  const [modalItem, setModalItem] = useState({});
  const { removeFromCart } = cartOperations;
  const cartItems = items.map((item) => (
    <CartItem
      key={item.id}
      item={item}
      language={language}
      user={user}
      setModalVisibility={setRemoveOneModalVisibility}
      setModalItem={setModalItem}
      cartOperations={cartOperations}
      promoCode={promoCode}
      certificateData={certificateData}
    />
  ));

  const onRemoveOneModalAction = (action) => {
    if (action) {
      removeFromCart(modalItem);
    }
    setRemoveOneModalVisibility(false);
  };
  return (
    <div className={styles.root}>
      {removeOneModalVisibility && (
        <ConfirmDialog
          title={t('common.modalHeader')}
          message={t('cart.deleteItem')}
          isOpen={removeOneModalVisibility}
          onAction={onRemoveOneModalAction}
          confirmButtonText={t('common.buttons.confirm')}
          dismisButtonText={t('common.buttons.cancel')}
        />
      )}
      <PageTitle title={t('cart.titleFilled')} />
      <div className={styles.table}>
        <Table>
          <TableHead>
            <TableRow
              classes={{
                root: styles.tableHeader
              }}
            >
              <TableCell>{t('cart.product')}</TableCell>
              <TableCell>{t('cart.size')}</TableCell>
              <TableCell>{t('cart.price')}</TableCell>
              <TableCell>{t('cart.quantity')}</TableCell>
              <TableCell>{t('cart.toPay')}</TableCell>
              <TableCell>{t('cart.actions')}</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>{cartItems}</TableBody>
        </Table>
      </div>
    </div>
  );
};
export default OrderTable;
