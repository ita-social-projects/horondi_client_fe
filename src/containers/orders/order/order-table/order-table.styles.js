import { makeStyles } from '@material-ui/core/styles';

const title = {
  fontWeight: '600',
  fontSize: '48px',
  lineHeight: '65px'
};

export const useStyles = makeStyles(({ palette }) => ({
  titleWrapper: {
    ...title
  },
  root: {
    maxWidth: '1440px',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-around',
    alignItems: 'center'
  },
  table: {
    width: '100%',
    '@media (max-width: 750px)': {
      width: '600px',
      overflowX: 'auto'
    },
    '@media (max-width: 600px)': {
      width: '400px',
      overflowX: 'auto'
    },
    '@media (max-width: 400px)': {
      width: '300px',
      overflowX: 'auto'
    },
    '& td': {
      padding: '32px 0',
      lineHeight: '28px',
      fontSize: '20px',
      fontWeight: 400,
      textAlign: 'center'
    },
    '& td:first-child': {
      textAlign: 'left',
      '& >div': {
        marginLeft: '5px'
      },
      '& img': {
        width: '220px',
        height: '133px',
        '@media (max-width: 600px)': {
          width: 'auto',
          height: 'auto'
        }
      }
    }
  },
  tableHeader: {
    borderTopWidth: '1px',
    borderTopStyle: 'solid',
    borderTopColor: palette.cart.borderColor,
    '& >th': {
      padding: '10px 0',
      fontSize: '14px',
      fontWeight: 600,
      lineHeight: '20px',
      textAlign: 'center'
    },
    '@media (max-width: 425px)': {
      '& >th': {
        verticalAlign: 'text-top'
      }
    }
  }
}));
