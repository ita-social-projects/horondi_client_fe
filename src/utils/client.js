import { ApolloClient, gql, split, ApolloLink, InMemoryCache } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { setContext } from '@apollo/client/link/context';
import fetch from 'unfetch';
import { IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
import { createUploadLink } from 'apollo-upload-client/public';
import { getMainDefinition } from '@apollo/client/utilities';

import { AUTH_ERRORS, USER_TOKENS } from '../configs';
import { getFromLocalStorage, setToLocalStorage } from '../services/local-storage.service';
import refreshAuthToken from './regenerateAuthTokenPair';

const introspectionResult = require('../fragmentTypes');

const { ACCESS_TOKEN } = USER_TOKENS;
const { ACCESS_TOKEN_IS_NOT_VALID } = AUTH_ERRORS;
const fragmentMatcher = new IntrospectionFragmentMatcher({
  introspectionQueryResultData: introspectionResult
});

export const REACT_APP_API_URL =
  window.env && window.env.REACT_APP_API_URL
    ? window.env.REACT_APP_API_URL
    : process.env.REACT_APP_API_URL;

const wsLink = new WebSocketLink({
  uri: process.env.REACT_APP_WS_API_URL,
  options: {
    reconnect: true,
    connectionParams: {
      authToken: getFromLocalStorage(ACCESS_TOKEN)
    }
  }
});

const authLink = setContext((_, { headers }) => ({
  headers: {
    ...headers,
    token: getFromLocalStorage(ACCESS_TOKEN) || ''
  }
}));

const terminatingLink = split(
  ({ query }) => {
    const { kind, operation } = getMainDefinition(query);
    return kind === 'OperationDefinition' && operation === 'subscription';
  },
  wsLink,
  authLink.concat(createUploadLink({ uri: `${REACT_APP_API_URL}/graphql` }))
);

export const client = new ApolloClient({
  fetch,
  link: ApolloLink.from([terminatingLink]),
  cache: new InMemoryCache({
    addTypename: false,
    fragmentMatcher
  })
});

export const getItems = async (query, variables = {}) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const queryResult = await client.query({
      query: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (queryResult.data && Object.values(queryResult.data)[0]?.message) {
      const message = Object.values(queryResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await getItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return queryResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setItems = async (query, variables) => {
  try {
    const token = getFromLocalStorage(ACCESS_TOKEN);
    const mutationResult = await client.mutate({
      mutation: gql`
        ${query}
      `,
      variables,
      context: {
        headers: {
          token
        }
      },
      fetchPolicy: 'no-cache'
    });

    if (mutationResult.data && Object.values(mutationResult.data)[0]?.message) {
      const message = Object.values(mutationResult.data)[0]?.message;
      if (message === ACCESS_TOKEN_IS_NOT_VALID) {
        const tokenResult = await refreshAuthToken();
        if (tokenResult) {
          return await setItems(query, variables);
        }

        setToLocalStorage(ACCESS_TOKEN, null);
      }

      throw new Error(message);
    }
    return mutationResult;
  } catch (e) {
    throw new Error(e.message);
  }
};

export const setChatMail = (query, variables) =>
  client.mutate({
    mutation: gql`
      ${query}
    `,
    variables
  });
