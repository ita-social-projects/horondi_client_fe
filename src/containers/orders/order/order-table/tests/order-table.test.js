import React from 'react';
import { render, screen, act, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import { MockedProvider } from '@apollo/client/testing';
import OrderTable from '../order-table';
import Modal from '../../../../../components/modal/modal';
import { mockGetProductById, props, modalProps, mockCartOperations } from './order-table.variables';

jest.mock('../order-table.styles', () => ({
  useStyles: () => ({})
}));
jest.mock('../../../../../components/modal/modal.styles', () => ({ useStyles: () => ({}) }));
jest.mock('../../../cart/cart-item/cart-item.styles', () => ({ useStyles: () => ({}) }));
jest.mock('react-redux');

const dispatch = jest.fn();
useDispatch.mockImplementation(() => dispatch);

const testUseSelector = (lang) => {
  useSelector.mockImplementation(() => ({
    currency: 0
  }));
};

describe('test <OrderTable /> component', () => {
  it('should render <OrderTable />', () => {
    testUseSelector(0);
    const { getByText } = render(
      <MockedProvider mocks={[mockGetProductById]} addTypename={false}>
        <OrderTable {...props} cartOperations={mockCartOperations} />
      </MockedProvider>
    );

    const productCell = getByText(/product/i);
    expect(productCell).toBeInTheDocument();
  });
});

describe('test <Modal/> component', () => {
  beforeEach(() => {
    render(<Modal {...modalProps} />);
  });

  it('should render <Modal /> component', () => {
    expect(screen.queryByText(modalProps.message)).toBeInTheDocument();
    expect(screen.queryByText('common.buttons.confirm')).toBeInTheDocument();
    expect(screen.queryByText('common.buttons.cancel')).toBeInTheDocument();
    expect(screen.queryByText('common.modalHeader')).toBeInTheDocument();
  });

  it('call onAction with true', () => {
    const confirmButton = screen.queryByText('common.buttons.confirm');
    act(() => {
      fireEvent.click(confirmButton);
    });

    expect(modalProps.onAction).toHaveBeenCalledWith(true);
  });

  it('call onAction with false', () => {
    const cancelButton = screen.queryByText('common.buttons.cancel');
    act(() => {
      fireEvent.click(cancelButton);
    });

    expect(modalProps.onAction).toHaveBeenCalledWith(false);
  });

  it('call onAction with false', () => {
    const closeModalIcon = screen.queryByTestId('closeModalIcon');
    act(() => {
      fireEvent.click(closeModalIcon);
    });

    expect(modalProps.onAction).toHaveBeenCalledWith(false);
  });
});
