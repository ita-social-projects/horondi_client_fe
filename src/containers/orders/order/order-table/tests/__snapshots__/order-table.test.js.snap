// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order table component tests should match snapshot 1`] = `
<div>
  <h2>
    cart.titleFilled
     
  </h2>
  <div>
    <WithStyles(ForwardRef(Table))>
      <WithStyles(ForwardRef(TableHead))>
        <WithStyles(ForwardRef(TableRow))
          classes={
            Object {
              "root": undefined,
            }
          }
        >
          <WithStyles(ForwardRef(TableCell))>
            cart.product
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))>
            cart.size
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))>
            cart.price
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))>
            cart.quantity
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))>
            cart.toPay
          </WithStyles(ForwardRef(TableCell))>
          <WithStyles(ForwardRef(TableCell))>
            cart.actions
          </WithStyles(ForwardRef(TableCell))>
        </WithStyles(ForwardRef(TableRow))>
      </WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableBody))>
        <CartItem
          calcPrice={100}
          cartLoading={false}
          cartQuantityLoading={false}
          currency={0}
          initialPrice={10}
          item={
            Object {
              "price": 10,
              "product": Object {
                "product_id": "some id",
              },
            }
          }
          language={1}
          setModalItem={[Function]}
          setModalVisibility={[Function]}
          user={Object {}}
        />
      </WithStyles(ForwardRef(TableBody))>
    </WithStyles(ForwardRef(Table))>
  </div>
</div>
`;
